### PR TITLE
AL/MIN-491. Prepare storage for liquidation.

### PR DESCRIPTION
**Pull request by lut1k**
**Description:**

1. Completely removed the implementation of the risk-manager.
2. Implemented storage in pallet risk-manager:
```
pub(crate) type LiquidationFee<T: Config> = StorageMap<_, Twox64Concat, CurrencyId, Rate, ValueQuery>;

pub(crate) type LiquidationThreshold<T: Config> = StorageValue<_, Rate, ValueQuery>;

pub(crate) type UserLiquidationAttempts<T: Config> = StorageMap<_, Twox64Concat, T::AccountId, u8, ValueQuery>;
```
3. Implemented the parameters `PartialLiquidationMinSum` and `PartialLiquidationMaxAttempts` using constants, not storage.
4. Implemented extrinsics for setting parameters in the pallet risk manager. Covered these functions with tests.
5. Implemented the public API of the pallet risk-manager using traits:
 ```
pub trait RiskManagerStorageProvider {
	fn create_pool(pool_id: CurrencyId, liquidation_threshold: Rate, liquidation_fee: Rate) -> DispatchResult;
	fn remove_pool(pool_id: CurrencyId);
}
pub trait UserLiquidationAttemptsManager<AccountId> {
	fn get_user_liquidation_attempts(who: &AccountId) -> u8;
	fn increase_by_one(who: &AccountId);
	fn reset_to_zero(who: &AccountId);
	fn mutate_depending_operation(pool_id: CurrencyId, who: &AccountId, operation: Operation);
}
```

7. Recalculated the weights for the minterest-protocol pallet extrinsics.

**Changes in storage or API:**

Changed types in types.json file:
Changed:
`PoolInitData : "threshold": "Rate" -> "liquidation_threshold": "Rate"`

Changed strucutre:
```
"PoolUserData": {
      "borrowed": "Balance",
      "interest_index": "Rate",
      "is_collateral": "bool"
}
```
Removed structure:
```
 "RiskManagerData": {
      "max_attempts": "u8",
      "min_partial_liquidation_sum": "Balance",
      "threshold": "Rate",
      "liquidation_fee": "Rate"
},
```

**Checklist**

- [x] Code is covered with tests
- [x] Code is covered with benchmarks
- [x] Fresh benchmark results included

### CODE REVIEW

File: pallets/risk-manager/src/lib.rs [hunk 1]
```diff
@@ -2,180 +2,103 @@
...
-	/// The additional collateral which is taken from borrowers as a penalty for being liquidated.
-	pub liquidation_fee: Rate,
-}
-
-type MinterestProtocol<T> = minterest_protocol::Pallet<T>;
-
 #[frame_support::pallet]
 pub mod module {
 	use super::*;
+	use frame_system::pallet_prelude::OriginFor;
```
Comment (Apollo50) [inline pallets/risk-manager/src/lib.rs:30]: plz move it outside

File: pallets/risk-manager/src/lib.rs [hunk 2]
```diff
@@ -314,437 +137,93 @@ pub mod module {
...
-			.map(|x| x.into_inner())
-			.ok_or(Error::<T>::NumOverflow)?;
-
-		Ok((seize_amount, is_attempt_increment_required))
-	}
-
+	/// Checks if liquidation_fee <= 0.5
 	fn is_valid_liquidation_fee(liquidation_fee: Rate) -> bool {
-		liquidation_fee >= Rate::one() && liquidation_fee <= Rate::saturating_from_rational(15, 10)
+		liquidation_fee <= Rate::saturating_from_rational(5, 10)
```
Comment (Apollo50) [inline pallets/risk-manager/src/lib.rs:182]: What do you think about moving this magic number to constants ?

File: service/src/chain_spec.rs [hunk 3]
```diff
@@ -385,44 +384,13 @@ fn minterest_genesis(
...
-					BTC,
-					RiskManagerData {
-						max_attempts: 2,
-						min_partial_liquidation_sum: 200_000 * DOLLARS, // In USD. FIXME: temporary value.
-						threshold: FixedU128::saturating_from_rational(103, 100), // 3%
-						liquidation_fee: FixedU128::saturating_from_rational(105, 100), // 5%
-					},
-				),
+			liquidation_fee: vec![
+				(DOT, FixedU128::saturating_from_rational(105, 100)),
```
Comment (Apollo50) [inline service/src/chain_spec.rs:388]:
```suggestion
				(DOT, FixedU128::saturating_from_rational(5, 100)),
```

File: test-helper/src/lib.rs [hunk 4]
```diff
@@ -268,21 +269,16 @@ macro_rules! mock_impl_minterest_protocol_config {
 macro_rules! mock_impl_risk_manager_config {
 	($target:ty, $acc:ident) => {
 		parameter_types! {
-			pub const RiskManagerPriority: TransactionPriority = TransactionPriority::max_value();
-			pub const MaxDurationMs: u64 = 1000;
+			pub const PartialLiquidationMinSum: Balance = 100_000 * DOLLARS;
```
Comment (Apollo50) [inline test-helper/src/lib.rs:272]: Maybe better do it throught the constants?

File: pallets/risk-manager/src/lib.rs (hunk 1 above)
Comment (lut1k) [reply pallets/risk-manager/src/lib.rs:30]: Fixed

File: pallets/risk-manager/src/lib.rs (hunk 2 above)
Comment (lut1k) [reply pallets/risk-manager/src/lib.rs:182]: Implement via constant.

File: service/src/chain_spec.rs (hunk 3 above)
Comment (lut1k) [reply service/src/chain_spec.rs:388]: Fixed

File: test-helper/src/lib.rs (hunk 4 above)
Comment (lut1k) [reply test-helper/src/lib.rs:272]: Left current implementation.

**Review by Apollo50 — DISMISSED** [review]

File: pallets/traits/src/lib.rs [hunk 5]
```diff
@@ -372,3 +351,31 @@ pub trait CurrencyConverter {
...
+	/// Increases the parameter liquidation_attempts by one for user.
+	fn increase_by_one(who: &AccountId);
+
+	/// Resets the parameter liquidation_attempts equal to zero for user.
+	fn reset_to_zero(who: &AccountId);
+
+	/// Mutates user liquidation attempts depending on user operation.
+	/// If the user makes a deposit to the collateral pool, then attempts are set to zero.
+	/// TODO: a liquidation handler will be added in the future
+	fn mutate_depending_operation(pool_id: CurrencyId, who: &AccountId, operation: Operation);
```
Comment (kurotych) [inline pallets/traits/src/lib.rs:369]: When we have `mutate_depending_operation`, Do we really need make `increase_by_one` and `reset_to_zero` public functions?

File: types.json [hunk 6]
```diff
@@ -1,204 +1,197 @@
...
+      "reserved": "Balance",
+      "misc_frozen": "Balance",
+      "fee_frozen": "Balance"
+    },
+    "OrmlAccountData": {
+      "free": "Balance",
+      "reserved": "Balance",
+      "frozen": "Balance"
+    },
+    "AccountData": "OrmlAccountData"
```
Comment (kurotych) [inline types.json:194]: Remove this formatting please.

Comment (lut1k) [reply types.json:194]: Fixind using four spaces

File: pallets/traits/src/lib.rs (hunk 5 above)
Comment (lut1k) [reply pallets/traits/src/lib.rs:369]: Made private

**Review by Jekshmek — DISMISSED** [review]

**Review by Apollo50 — DISMISSED** [review]

**Review by kurotych — APPROVED** [review]

**Review by Apollo50 — APPROVED** [review]

**Pull request merged**